README.md
# Ember-spinner-button

Simple Ember spinner button component using [spin.js](http://fgnass.github.io/spin.js/).

[![Build Status](https://travis-ci.org/RSSchermer/ember-spinner-button.svg?branch=master)](https://travis-ci.org/RSSchermer/ember-spinner-button)

## Demo
Demo available [here](http://rsschermer.github.io/ember-spinner-button/).

## Installation

``` bash
ember install ember-spinner-button
```

## Usage

Example:

``` handlebars
{{#spinner-button onclick=(action "saveForm") isSpinning=isSaving}}
  Save
{{/spinner-button}}
```

The following properties should be set:

* `isSpinning`: a boolean value indicating whether or not the button is currently spinning.
* `onclick`: the action that will be called when the button is clicked while not currently spinning.

The following properties can be set to customize the button:

* `disabled` (default: `false`): boolean value or function that specifies the `disabled` attribute.
* `title` (default: `null`): string or function that resolves to the `title` attribute.

The following properties can be set to customize the spinner on a particular button:

* `lines` (default: `8`): the number of lines to draw.
* `length` (default: `4`): the length of each line.
* `width` (default: `3`): the line thickness.
* `radius` (default: `5`): the radius of the inner circle.
* `corners` (default: `1`): corner roundness (0..1).
* `rotate` (default: `0`): the rotation offset.
* `direction` (default: `1`): 1 for clockwise, -1 for counterclockwise.
* `color` (default: parent element text color): #rgb or #rrggbb or array of colors.
* `speed` (default: `1`): rounds per second.
* `trail` (default: `60`): afterglow percentage.
* `shadow` (default: `false`): whether to render a shadow.
* `hwaccel` (default: `false`): whether to use hardware acceleration.
* `top` (default: `'50%'`): top position relative to parent.
* `left` (default: `'50%'`): left position relative to parent.

If you want to change the defaults for all spinner buttons in your application, extend the component with your own
defaults:

```javascript
// app/components/spinner-button.js
import SpinnerButtonComponent from 'ember-spinner-button/components/spinner-button';

export default SpinnerButtonComponent.extend({
  length: 6,

  width: 2
});
```


CHANGELOG.md
# Ember spinner-button change log

## 0.2.0

An action can now be bound that the `onclick` attribute. This action will be invoked when the button is clicked:

``` handlebars
{{#spinner-button onclick=(action "saveForm") isSpinning=isSaving}}
  Save
{{/spinner-button}}
```

This should be preferred over binding a string to the `action` attribute.

## 0.1.0

Adds the following properties that can be set to customize the button:

* `disabled` (default: `false`): boolean value or function that specifies the `disabled` attribute.
* `title` (default: `null`): string or function that resolves to the `title` attribute.


tests/dummy/app/controllers/application.js
import Ember from 'ember';

export default Ember.Controller.extend({
  isBusy: false,

  actions: {
    makeBusy() {
      this.set('isBusy', true);
    }
  }
});


tests/unit/components/spinner-button-test.js
import { test, moduleForComponent } from 'ember-qunit';
import Ember from 'ember';

moduleForComponent('spinner-button', 'SpinnerButtonComponent', {
  needs: ['component:spin-spinner']
});

test('does not display a spinner when isSpinning is false', function (assert) {
  let $component = this.append();

  assert.equal($component.find('.spin-spinner').length, 0);
});

test('does display a spinner when isSpinning is true', function (assert) {
  let component = this.subject();
  let $component = this.append();

  Ember.run(() => {
    component.set('isSpinning', true);
  });

  assert.equal($component.find('.spin-spinner').length, 1);
});

test('sends the action when clicked when isSpinning is false', function (assert) {
  let component = this.subject();
  let $component = this.append();

  let targetObject = {
    dummyAction: () => {
      assert.ok(true, 'external Action was called!');
    }
  };

  Ember.run(() => {
    component.setProperties({ 'action': 'dummyAction', 'targetObject': targetObject });
  });

  Ember.run(() => {
    $component.click();
  });
});


tests/integration/spinner-button-component-test.js
import hbs from 'htmlbars-inline-precompile';
import { test, moduleForComponent } from 'ember-qunit';

moduleForComponent('spinner-button', 'SpinnerButtonComponent', {
  integration: true
});

test('does not display a spinner when isSpinning is false', function (assert) {
  this.render(hbs`
    {{#spinner-button isSpinning=false}}
      Click me
    {{/spinner-button}}
  `);

  assert.equal(this.$('.spin-spinner').length, 0);
});

test('does display a spinner when isSpinning is true', function (assert) {
  this.render(hbs`
    {{#spinner-button isSpinning=true}}
      Click me
    {{/spinner-button}}
  `);

  assert.equal(this.$('.spin-spinner').length, 1);
});

test('sends the action when clicked when isSpinning is false', function (assert) {
  this.render(hbs`
    {{#spinner-button id="my_button" action='dummyAction' isSpinning=false}}
      Click me
    {{/spinner-button}}
  `);

  this.set('actions.dummyAction', () => {
    assert.ok(true, 'external Action was called!');
  });

  this.$('#my_button').click();
});

test('fires the action bound to onclick when clicked when isSpinning is false', function (assert) {
  this.set('actions.dummyAction', () => {
    assert.ok(true, 'external Action was called!');
  });
  
  this.render(hbs`
    {{#spinner-button id="my_button" onclick=(action 'dummyAction') isSpinning=false}}
      Click me
    {{/spinner-button}}
  `);

  this.$('#my_button').click();
});


demo-app/config/environment.js
/* jshint node: true */

module.exports = function(environment) {
  var ENV = {
    modulePrefix: 'spinner-button-demo-app',
    environment: environment,
    baseURL: '/ember-spinner-button',
    locationType: 'auto',
    contentSecurityPolicy: {
      'default-src': "'none'",
      'script-src': "'self' 'unsafe-eval'",
      'font-src': "'self'",
      'connect-src': "'self'",
      'img-src': "'self'",
      'style-src': "'self' 'unsafe-inline'",
      'media-src': "'self'"
    },
    EmberENV: {
      FEATURES: {
        // Here you can enable experimental features on an ember canary build
        // e.g. 'with-controller': true
      }
    },

    APP: {
      // Here you can pass flags/options to your application instance
      // when it is created
    }
  };

  if (environment === 'development') {
    // ENV.APP.LOG_RESOLVER = true;
    ENV.APP.LOG_ACTIVE_GENERATION = true;
    // ENV.APP.LOG_TRANSITIONS = true;
    // ENV.APP.LOG_TRANSITIONS_INTERNAL = true;
    ENV.APP.LOG_VIEW_LOOKUPS = true;
  }

  if (environment === 'test') {
    // Testem prefers this...
    ENV.baseURL = '/';
    ENV.locationType = 'auto';

    // keep test console output quieter
    ENV.APP.LOG_ACTIVE_GENERATION = false;
    ENV.APP.LOG_VIEW_LOOKUPS = false;

    ENV.APP.rootElement = '#ember-testing';
  }

  if (environment === 'production') {

  }

  return ENV;
};


assets/spinner-button-demo-app.js
eval("//# sourceURL=vendor/ember-cli/loader.js");

;eval("define(\"spinner-button-demo-app/app\", \n  [\"ember\",\"ember/resolver\",\"ember/load-initializers\",\"spinner-button-demo-app/config/environment\",\"exports\"],\n  function(__dependency1__, __dependency2__, __dependency3__, __dependency4__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n    var Resolver = __dependency2__[\"default\"];\n    var loadInitializers = __dependency3__[\"default\"];\n    var config = __dependency4__[\"default\"];\n\n    Ember.MODEL_FACTORY_INJECTIONS = true;\n\n    var App = Ember.Application.extend({\n      modulePrefix: config.modulePrefix,\n      podModulePrefix: config.podModulePrefix,\n      Resolver: Resolver\n    });\n\n    loadInitializers(App, config.modulePrefix);\n\n    __exports__[\"default\"] = App;\n  });//# sourceURL=spinner-button-demo-app/app.js");

;eval("define(\"spinner-button-demo-app/components/spin-spinner\", \n  [\"ember-spin-spinner/components/spin-spinner\",\"exports\"],\n  function(__dependency1__, __exports__) {\n    \"use strict\";\n    var SpinSpinnerComponent = __dependency1__[\"default\"];\n\n    \r\n    __exports__[\"default\"] = SpinSpinnerComponent;\n  });//# sourceURL=spinner-button-demo-app/components/spin-spinner.js");

;eval("define(\"spinner-button-demo-app/components/spinner-button\", \n  [\"ember-spinner-button/components/spinner-button\",\"exports\"],\n  function(__dependency1__, __exports__) {\n    \"use strict\";\n    var SpinnerButtonComponent = __dependency1__[\"default\"];\n\n    __exports__[\"default\"] = SpinnerButtonComponent;\n  });//# sourceURL=spinner-button-demo-app/components/spinner-button.js");

;eval("define(\"spinner-button-demo-app/controllers/application\", \n  [\"ember\",\"exports\"],\n  function(__dependency1__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n\n    __exports__[\"default\"] = Ember.Controller.extend({\n      isBusy: false,\n\n      actions: {\n        makeBusy: function () {\n          this.set(\'isBusy\', true);\n        }\n      }\n    });\n  });//# sourceURL=spinner-button-demo-app/controllers/application.js");

;eval("define(\"spinner-button-demo-app/router\", \n  [\"ember\",\"spinner-button-demo-app/config/environment\",\"exports\"],\n  function(__dependency1__, __dependency2__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n    var config = __dependency2__[\"default\"];\n\n    var Router = Ember.Router.extend({\n      location: config.locationType\n    });\n\n    Router.map(function() {\n    });\n\n    __exports__[\"default\"] = Router;\n  });//# sourceURL=spinner-button-demo-app/router.js");

;eval("define(\"spinner-button-demo-app/templates/application\", \n  [\"ember\",\"exports\"],\n  function(__dependency1__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n    __exports__[\"default\"] = Ember.Handlebars.template(function anonymous(Handlebars,depth0,helpers,partials,data) {\n    this.compilerInfo = [4,\'>= 1.0.0\'];\n    helpers = this.merge(helpers, Ember.Handlebars.helpers); data = data || {};\n      var buffer = \'\', stack1, helper, options, helperMissing=helpers.helperMissing, escapeExpression=this.escapeExpression, self=this;\n\n    function program1(depth0,data) {\n      \n      \n      data.buffer.push(\"\\n      I\'m a spinner-button!\\n    \");\n      }\n\n      data.buffer.push(\"<section class=\\\"container\\\">\\n  <h1 id=\'title\'>Ember spinner-button Demo</h2>\\n\\n  <p>This is a demo application for the Ember spinner-button component. Documentation for this component can be found\\n    <a href=\\\"https://github.com/RSSchermer/ember-spinner-button\\\">here</a>. The source for this demo application can be found\\n    <a href=\\\"https://github.com/RSSchermer/ember-spinner-button/tree/gh-pages/demo-app\\\">here</a>.\\n  </p>\\n\\n  <p>This demo uses Twitter Bootstrap, which affects the appearence of the spinner-button as you would expect.</p>\\n\\n  <p>This isSpinning property of this button is bound to the isBusy property on the application controller. You can use\\n    the checkbox to toggle the value of isBusy. Clicking the button sends a \'makeBusy\' event to the controller. The\\n    handler for this event on the controller sets isBusy to true.\\n  </p>\\n\\n  <p>\\n    <label>\\n      \");\n      data.buffer.push(escapeExpression((helper = helpers.input || (depth0 && depth0.input),options={hash:{\n        \'type\': (\"checkbox\"),\n        \'name\': (\"isBusy\"),\n        \'checked\': (\"isBusy\")\n      },hashTypes:{\'type\': \"STRING\",\'name\': \"STRING\",\'checked\': \"ID\"},hashContexts:{\'type\': depth0,\'name\': depth0,\'checked\': depth0},contexts:[],types:[],data:data},helper ? helper.call(depth0, options) : helperMissing.call(depth0, \"input\", options))));\n      data.buffer.push(\"\\n      isBusy\\n    </label>\\n  </p>\\n\\n  <p>\\n    \");\n      stack1 = (helper = helpers[\'spinner-button\'] || (depth0 && depth0[\'spinner-button\']),options={hash:{\n        \'class\': (\"btn btn-primary\"),\n        \'action\': (\"makeBusy\"),\n        \'isSpinning\': (\"isBusy\")\n      },hashTypes:{\'class\': \"STRING\",\'action\': \"STRING\",\'isSpinning\': \"ID\"},hashContexts:{\'class\': depth0,\'action\': depth0,\'isSpinning\': depth0},inverse:self.noop,fn:self.program(1, program1, data),contexts:[],types:[],data:data},helper ? helper.call(depth0, options) : helperMissing.call(depth0, \"spinner-button\", options));\n      if(stack1 || stack1 === 0) { data.buffer.push(stack1); }\n      data.buffer.push(\"\\n  </p>\\n</section>\\n\");\n      return buffer;\n      \n    });\n  });//# sourceURL=spinner-button-demo-app/templates/application.js");

;eval("define(\"spinner-button-demo-app/templates/components/spinner-button\", \n  [\"ember\",\"exports\"],\n  function(__dependency1__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n    __exports__[\"default\"] = Ember.Handlebars.template(function anonymous(Handlebars,depth0,helpers,partials,data) {\n    this.compilerInfo = [4,\'>= 1.0.0\'];\n    helpers = this.merge(helpers, Ember.Handlebars.helpers); data = data || {};\n      var buffer = \'\', stack1, helperMissing=helpers.helperMissing, escapeExpression=this.escapeExpression, self=this;\n\n    function program1(depth0,data) {\n      \n      var buffer = \'\', helper, options;\n      data.buffer.push(\"\\n  \");\n      data.buffer.push(escapeExpression((helper = helpers[\'spin-spinner\'] || (depth0 && depth0[\'spin-spinner\']),options={hash:{\n        \'lines\': (\"lines\"),\n        \'length\': (\"length\"),\n        \'width\': (\"width\"),\n        \'radius\': (\"radius\"),\n        \'corners\': (\"corners\"),\n        \'rotate\': (\"rotate\"),\n        \'direction\': (\"direction\"),\n        \'speed\': (\"speed\"),\n        \'trail\': (\"trail\"),\n        \'shadow\': (\"shadow\"),\n        \'hwaccel\': (\"hwaccel\"),\n        \'color\': (\"color\"),\n        \'left\': (\"left\"),\n        \'top\': (\"top\")\n      },hashTypes:{\'lines\': \"ID\",\'length\': \"ID\",\'width\': \"ID\",\'radius\': \"ID\",\'corners\': \"ID\",\'rotate\': \"ID\",\'direction\': \"ID\",\'speed\': \"ID\",\'trail\': \"ID\",\'shadow\': \"ID\",\'hwaccel\': \"ID\",\'color\': \"ID\",\'left\': \"ID\",\'top\': \"ID\"},hashContexts:{\'lines\': depth0,\'length\': depth0,\'width\': depth0,\'radius\': depth0,\'corners\': depth0,\'rotate\': depth0,\'direction\': depth0,\'speed\': depth0,\'trail\': depth0,\'shadow\': depth0,\'hwaccel\': depth0,\'color\': depth0,\'left\': depth0,\'top\': depth0},contexts:[],types:[],data:data},helper ? helper.call(depth0, options) : helperMissing.call(depth0, \"spin-spinner\", options))));\n      data.buffer.push(\"\\n\");\n      return buffer;\n      }\n\n    function program3(depth0,data) {\n      \n      var buffer = \'\', stack1;\n      data.buffer.push(\"\\n  \");\n      stack1 = helpers._triageMustache.call(depth0, \"yield\", {hash:{},hashTypes:{},hashContexts:{},contexts:[depth0],types:[\"ID\"],data:data});\n      if(stack1 || stack1 === 0) { data.buffer.push(stack1); }\n      data.buffer.push(\"\\n\");\n      return buffer;\n      }\n\n      stack1 = helpers[\'if\'].call(depth0, \"isSpinning\", {hash:{},hashTypes:{},hashContexts:{},inverse:self.program(3, program3, data),fn:self.program(1, program1, data),contexts:[depth0],types:[\"ID\"],data:data});\n      if(stack1 || stack1 === 0) { data.buffer.push(stack1); }\n      data.buffer.push(\"\\n\");\n      return buffer;\n      \n    });\n  });//# sourceURL=spinner-button-demo-app/templates/components/spinner-button.js");

;eval("define(\"spinner-button-demo-app/tests/app.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - .\');\n    test(\'app.js should pass jshint\', function() { \n      ok(true, \'app.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/app.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/controllers/application.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - controllers\');\n    test(\'controllers/application.js should pass jshint\', function() { \n      ok(true, \'controllers/application.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/controllers/application.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/helpers/resolver\", \n  [\"ember/resolver\",\"spinner-button-demo-app/config/environment\",\"exports\"],\n  function(__dependency1__, __dependency2__, __exports__) {\n    \"use strict\";\n    var Resolver = __dependency1__[\"default\"];\n    var config = __dependency2__[\"default\"];\n\n    var resolver = Resolver.create();\n\n    resolver.namespace = {\n      modulePrefix: config.modulePrefix,\n      podModulePrefix: config.podModulePrefix\n    };\n\n    __exports__[\"default\"] = resolver;\n  });//# sourceURL=spinner-button-demo-app/tests/helpers/resolver.js");

;eval("define(\"spinner-button-demo-app/tests/helpers/start-app\", \n  [\"ember\",\"spinner-button-demo-app/app\",\"spinner-button-demo-app/router\",\"spinner-button-demo-app/config/environment\",\"exports\"],\n  function(__dependency1__, __dependency2__, __dependency3__, __dependency4__, __exports__) {\n    \"use strict\";\n    var Ember = __dependency1__[\"default\"];\n    var Application = __dependency2__[\"default\"];\n    var Router = __dependency3__[\"default\"];\n    var config = __dependency4__[\"default\"];\n\n    __exports__[\"default\"] = function startApp(attrs) {\n      var App;\n\n      var attributes = Ember.merge({}, config.APP);\n      attributes = Ember.merge(attributes, attrs); // use defaults, but you can override;\n\n      Router.reopen({\n        location: \'none\'\n      });\n\n      Ember.run(function() {\n        App = Application.create(attributes);\n        App.setupForTesting();\n        App.injectTestHelpers();\n      });\n\n      App.reset(); // this shouldn\'t be needed, i want to be able to \"start an app at a specific URL\"\n\n      return App;\n    }\n  });//# sourceURL=spinner-button-demo-app/tests/helpers/start-app.js");

;eval("define(\"spinner-button-demo-app/tests/router.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - .\');\n    test(\'router.js should pass jshint\', function() { \n      ok(true, \'router.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/router.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/spinner-button-demo-app/tests/helpers/resolver.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - spinner-button-demo-app/tests/helpers\');\n    test(\'spinner-button-demo-app/tests/helpers/resolver.js should pass jshint\', function() { \n      ok(true, \'spinner-button-demo-app/tests/helpers/resolver.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/spinner-button-demo-app/tests/helpers/resolver.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/spinner-button-demo-app/tests/helpers/start-app.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - spinner-button-demo-app/tests/helpers\');\n    test(\'spinner-button-demo-app/tests/helpers/start-app.js should pass jshint\', function() { \n      ok(true, \'spinner-button-demo-app/tests/helpers/start-app.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/spinner-button-demo-app/tests/helpers/start-app.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/spinner-button-demo-app/tests/test-helper.jshint\", \n  [],\n  function() {\n    \"use strict\";\n    module(\'JSHint - spinner-button-demo-app/tests\');\n    test(\'spinner-button-demo-app/tests/test-helper.js should pass jshint\', function() { \n      ok(true, \'spinner-button-demo-app/tests/test-helper.js should pass jshint.\'); \n    });\n  });//# sourceURL=spinner-button-demo-app/tests/spinner-button-demo-app/tests/test-helper.jshint.js");

;eval("define(\"spinner-button-demo-app/tests/test-helper\", \n  [\"spinner-button-demo-app/tests/helpers/resolver\",\"ember-qunit\"],\n  function(__dependency1__, __dependency2__) {\n    \"use strict\";\n    var resolver = __dependency1__[\"default\"];\n    var setResolver = __dependency2__.setResolver;\n\n    setResolver(resolver);\n\n    document.write(\'<div id=\"ember-testing-container\"><div id=\"ember-testing\"></div></div>\');\n\n    QUnit.config.urlConfig.push({ id: \'nocontainer\', label: \'Hide container\'});\n    var containerVisibility = QUnit.urlParams.nocontainer ? \'hidden\' : \'visible\';\n    document.getElementById(\'ember-testing-container\').style.visibility = containerVisibility;\n  });//# sourceURL=spinner-button-demo-app/tests/test-helper.js");

/* jshint ignore:start */

define('spinner-button-demo-app/config/environment', ['ember'], function(Ember) {
  var metaName = 'spinner-button-demo-app/config/environment';
  var rawConfig = Ember['default'].$('meta[name="' + metaName + '"]').attr('content');
  var config = JSON.parse(unescape(rawConfig));

  return { 'default': config };
});

if (runningTests) {
  require('spinner-button-demo-app/tests/test-helper');
} else {
  require('spinner-button-demo-app/app')['default'].create({"LOG_ACTIVE_GENERATION":true,"LOG_VIEW_LOOKUPS":true});
}

/* jshint ignore:end */


addon/components/spinner-button.js
import Ember from 'ember';

export default Ember.Component.extend({
  tagName: 'button',

  classNames: ['spinner-button'],

  classNameBindings: ['spinning'],

  attributeBindings: [ 'disabled', 'title'],

  disabled: false,

  title: null,

  isSpinning: false,

  action: null,

  onclick: null,

  lines: 8,

  radius: 5,

  length: 4,

  width: 3,

  corners: 1,

  rotate: 0,

  direction: 1,

  speed: 1,

  trail: 60,

  shadow: false,

  hwaccel: false,

  color: null,

  left: '50%',

  top: '50%',

  setPositionToRelative: Ember.on('didInsertElement', function () {
    this.$().css('position', 'relative');
  }),

  maintainButtonDimensions: Ember.on('didInsertElement', Ember.observer('isSpinning', function () {
    if (this.get('isSpinning')) {
      this.$().css({
        'width': this.$().outerWidth() +'px',
        'height': this.$().outerHeight() +'px'
      });
    } else {
      this.$().css({
        'width': '',
        'height': ''
      });
    }
  })),

  click() {
    if (!this.get('isSpinning')) {
      if (this.get('action')) {
        this.sendAction();
      }

      if (typeof this.get('onclick') === 'function') {
        this.get('onclick')();
      }
    }
  }
});
